File: src/types/unified.ts
// Unified Data Architecture for SiYuan + Heptabase Integration
// This file defines the core types for a unified block-based system that supports
// both document hierarchy (SiYuan-style) and spatial whiteboard representation (Heptabase-style)

import type { Position, Size, NodeStyle, ConnectionStyle } from './index'

// ============================================================================
// CORE UNIFIED BLOCK TYPES
// ============================================================================

/**
 * Enhanced block type that supports both traditional document blocks
 * and spatial whiteboard elements as a unified entity
 */
export interface UnifiedBlock {
  id: string
  type: UnifiedBlockType
  content: string
  properties: Record<string, any>
  children: string[] // Child block IDs for hierarchical structure
  parentId?: string
  position: number // Position in document hierarchy
  
  // Spatial properties (optional - only for blocks that appear on whiteboard)
  spatialProperties?: SpatialProperties
  
  // Metadata
  metadata: UnifiedBlockMetadata
}

/**
 * Extended block types that support both document and spatial representations
 */
export type UnifiedBlockType = 
  // Traditional document blocks
  | 'paragraph' 
  | 'heading' 
  | 'list' 
  | 'code' 
  | 'quote' 
  | 'image' 
  | 'table'
  | 'embed'
  // Spatial-first blocks (primarily for whiteboard)
  | 'whiteboard-text'
  | 'whiteboard-image' 
  | 'whiteboard-group'
  | 'mindmap-node'
  | 'shape'
  | 'sticky-note'

/**
 * Spatial properties that allow any block to appear on whiteboards
 */
export interface SpatialProperties {
  canvasId: string // Which canvas this block appears on
  position: Position // X,Y coordinates on canvas
  size: Size
  style: SpatialNodeStyle
  connections: SpatialConnection[] // Connections to other blocks
  isVisible: boolean // Whether visible on current canvas
  layer: number // Z-index for layering
  locked: boolean // Whether the spatial position is locked
}

/**
 * Enhanced node style for spatial representation
 */
export interface SpatialNodeStyle extends NodeStyle {
  // Additional spatial-specific styling
  shadow: boolean
  shadowColor: string
  shadowBlur: number
  rotation: number // Rotation angle in degrees
  skewX: number
  skewY: number
}

/**
 * Spatial connections between blocks on whiteboards
 */
export interface SpatialConnection {
  id: string
  targetBlockId: string
  type: SpatialConnectionType
  style: ConnectionStyle
  label?: string
  points?: Position[] // For custom connection paths
  metadata: ConnectionMetadata
}

export type SpatialConnectionType = 
  | 'line' 
  | 'arrow' 
  | 'curve' 
  | 'dashed'
  | 'bidirectional'
  | 'hierarchical' // Represents parent-child relationship
  | 'reference' // Represents block reference/link
  | 'semantic' // User-defined semantic relationship

export interface ConnectionMetadata {
  createdAt: Date
  updatedAt: Date
  semanticType?: string // User-defined connection meaning
  strength?: number // Connection strength (0-1)
}

/**
 * Enhanced metadata for unified blocks
 */
export interface UnifiedBlockMetadata {
  createdAt: Date
  updatedAt: Date
  tags: string[]
  locked: boolean
  
  // Document-specific metadata
  wordCount?: number
  readingTime?: number
  
  // Spatial-specific metadata
  lastCanvasId?: string // Last canvas this block appeared on
  spatialHistory?: SpatialHistoryEntry[] // History of spatial positions
  
  // Cross-reference metadata
  references: BlockReference[] // Other blocks that reference this one
  backlinks: string[] // Block IDs that this block references
}

export interface SpatialHistoryEntry {
  canvasId: string
  position: Position
  timestamp: Date
}

export interface BlockReference {
  sourceBlockId: string
  referenceType: 'embed' | 'link' | 'mention'
  context?: string // Surrounding text context
}

// ============================================================================
// CONTAINER TYPES
// ============================================================================

/**
 * Document container that organizes blocks hierarchically
 */
export interface DocumentContainer {
  id: string
  title: string
  rootBlockIds: string[] // Top-level blocks in this document
  metadata: DocumentContainerMetadata
  path: string
}

export interface DocumentContainerMetadata {
  tags: string[]
  category: string
  author: string
  createdAt: Date
  updatedAt: Date
  wordCount: number
  readingTime: number
  lastModified: Date
}

/**
 * Canvas container that organizes blocks spatially
 */
export interface CanvasContainer {
  id: string
  name: string
  blockIds: string[] // Blocks that appear on this canvas
  viewport: Viewport
  settings: CanvasSettings
  metadata: CanvasContainerMetadata
}

export interface Viewport {
  x: number
  y: number
  zoom: number
}

export interface CanvasSettings {
  gridEnabled: boolean
  snapToGrid: boolean
  gridSize: number
  backgroundColor: string
  showMinimap: boolean
  allowOverlap: boolean
  autoLayout: boolean
}

export interface CanvasContainerMetadata {
  createdAt: Date
  updatedAt: Date
  tags: string[]
  description?: string
  thumbnail?: string
}

// ============================================================================
// WORKSPACE AND QUERY TYPES
// ============================================================================

/**
 * Workspace that contains all unified data
 */
export interface UnifiedWorkspace {
  id: string
  name: string
  blocks: Map<string, UnifiedBlock>
  documents: Map<string, DocumentContainer>
  canvases: Map<string, CanvasContainer>
  metadata: WorkspaceMetadata
}

export interface WorkspaceMetadata {
  createdAt: Date
  updatedAt: Date
  version: string
  lastSyncTime?: Date
  totalBlocks: number
  totalDocuments: number
  totalCanvases: number
}

/**
 * Query interface for advanced block operations
 */
export interface BlockQuery {
  // Content filters
  contentContains?: string
  contentRegex?: string
  
  // Type filters
  types?: UnifiedBlockType[]
  excludeTypes?: UnifiedBlockType[]
  
  // Hierarchical filters
  hasParent?: boolean
  hasChildren?: boolean
  parentId?: string
  ancestorId?: string
  depth?: number
  
  // Spatial filters
  onCanvas?: string
  inRegion?: {
    topLeft: Position
    bottomRight: Position
  }
  connectedTo?: string
  
  // Metadata filters
  tags?: string[]
  createdAfter?: Date
  createdBefore?: Date
  updatedAfter?: Date
  updatedBefore?: Date
  
  // Sorting and pagination
  sortBy?: 'createdAt' | 'updatedAt' | 'position' | 'content'
  sortOrder?: 'asc' | 'desc'
  limit?: number
  offset?: number
}

/**
 * Options for creating new blocks
 */
export interface CreateBlockOptions {
  parentId?: string
  position?: number
  insertAfter?: string
  
  // Spatial options
  addToCanvas?: string
  spatialPosition?: Position
  spatialSize?: Size
  
  // Initial properties
  properties?: Record<string, any>
  tags?: string[]
}

/**
 * Options for updating blocks
 */
export interface UpdateBlockOptions {
  // Content updates
  content?: string
  type?: UnifiedBlockType
  properties?: Record<string, any>
  
  // Hierarchical updates
  newParentId?: string
  newPosition?: number
  
  // Spatial updates
  spatialProperties?: Partial<SpatialProperties>
  
  // Metadata updates
  tags?: string[]
  addTags?: string[]
  removeTags?: string[]
}

// ============================================================================
// OPERATION RESULT TYPES
// ============================================================================

export interface BlockOperationResult {
  success: boolean
  blockId?: string
  error?: string
  affectedBlocks?: string[]
}

export interface QueryResult {
  blocks: UnifiedBlock[]
  totalCount: number
  hasMore: boolean
}

// ============================================================================
// EVENT TYPES FOR REAL-TIME SYNC
// ============================================================================

export interface UnifiedSyncEvent {
  id: string
  type: UnifiedSyncEventType
  blockId: string
  data: any
  timestamp: Date
  userId?: string
}

export type UnifiedSyncEventType = 
  | 'block_created'
  | 'block_updated' 
  | 'block_deleted'
  | 'block_moved'
  | 'spatial_updated'
  | 'connection_created'
  | 'connection_deleted'
  | 'canvas_updated'
  | 'document_updated'

File: src-tauri/src/unified.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// ============================================================================
// UNIFIED BLOCK TYPES
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UnifiedBlock {
    pub id: String,
    pub block_type: String,
    pub content: String,
    pub properties: HashMap<String, serde_json::Value>,
    pub children: Vec<String>,
    pub parent_id: Option<String>,
    pub position: u32,
    pub spatial_properties: Option<SpatialProperties>,
    pub metadata: UnifiedBlockMetadata,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpatialProperties {
    pub canvas_id: String,
    pub position: Position,
    pub size: Size,
    pub style: SpatialNodeStyle,
    pub connections: Vec<SpatialConnection>,
    pub is_visible: bool,
    pub layer: i32,
    pub locked: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Position {
    pub x: f64,
    pub y: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Size {
    pub width: f64,
    pub height: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpatialNodeStyle {
    pub background_color: String,
    pub border_color: String,
    pub border_width: f64,
    pub border_radius: f64,
    pub font_size: f64,
    pub font_color: String,
    pub opacity: f64,
    pub shadow: bool,
    pub shadow_color: String,
    pub shadow_blur: f64,
    pub rotation: f64,
    pub skew_x: f64,
    pub skew_y: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpatialConnection {
    pub id: String,
    pub target_block_id: String,
    pub connection_type: String,
    pub style: ConnectionStyle,
    pub label: Option<String>,
    pub points: Option<Vec<Position>>,
    pub metadata: ConnectionMetadata,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectionStyle {
    pub color: String,
    pub width: f64,
    pub opacity: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectionMetadata {
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
    pub semantic_type: Option<String>,
    pub strength: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UnifiedBlockMetadata {
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
    pub tags: Vec<String>,
    pub locked: bool,
    pub word_count: Option<u32>,
    pub reading_time: Option<u32>,
    pub last_canvas_id: Option<String>,
    pub spatial_history: Vec<SpatialHistoryEntry>,
    pub references: Vec<BlockReference>,
    pub backlinks: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpatialHistoryEntry {
    pub canvas_id: String,
    pub position: Position,
    pub timestamp: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlockReference {
    pub source_block_id: String,
    pub reference_type: String,
    pub context: Option<String>,
}

// ============================================================================
// CONTAINER TYPES
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DocumentContainer {
    pub id: String,
    pub title: String,
    pub root_block_ids: Vec<String>,
    pub metadata: DocumentContainerMetadata,
    pub path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DocumentContainerMetadata {
    pub tags: Vec<String>,
    pub category: String,
    pub author: String,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
    pub word_count: u32,
    pub reading_time: u32,
    pub last_modified: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CanvasContainer {
    pub id: String,
    pub name: String,
    pub block_ids: Vec<String>,
    pub viewport: Viewport,
    pub settings: CanvasSettings,
    pub metadata: CanvasContainerMetadata,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Viewport {
    pub x: f64,
    pub y: f64,
    pub zoom: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CanvasSettings {
    pub grid_enabled: bool,
    pub snap_to_grid: bool,
    pub grid_size: f64,
    pub background_color: String,
    pub show_minimap: bool,
    pub allow_overlap: bool,
    pub auto_layout: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CanvasContainerMetadata {
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
    pub tags: Vec<String>,
    pub description: Option<String>,
    pub thumbnail: Option<String>,
}

// ============================================================================
// WORKSPACE TYPES
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UnifiedWorkspace {
    pub id: String,
    pub name: String,
    pub blocks: HashMap<String, UnifiedBlock>,
    pub documents: HashMap<String, DocumentContainer>,
    pub canvases: HashMap<String, CanvasContainer>,
    pub metadata: WorkspaceMetadata,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorkspaceMetadata {
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
    pub version: String,
    pub last_sync_time: Option<chrono::DateTime<chrono::Utc>>,
    pub total_blocks: u32,
    pub total_documents: u32,
    pub total_canvases: u32,
}

// ============================================================================
// QUERY TYPES
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlockQuery {
    // Content filters
    pub content_contains: Option<String>,
    pub content_regex: Option<String>,
    
    // Type filters
    pub types: Option<Vec<String>>,
    pub exclude_types: Option<Vec<String>>,
    
    // Hierarchical filters
    pub has_parent: Option<bool>,
    pub has_children: Option<bool>,
    pub parent_id: Option<String>,
    pub ancestor_id: Option<String>,
    pub depth: Option<u32>,
    
    // Spatial filters
    pub on_canvas: Option<String>,
    pub in_region: Option<Region>,
    pub connected_to: Option<String>,
    
    // Metadata filters
    pub tags: Option<Vec<String>>,
    pub created_after: Option<chrono::DateTime<chrono::Utc>>,
    pub created_before: Option<chrono::DateTime<chrono::Utc>>,
    pub updated_after: Option<chrono::DateTime<chrono::Utc>>,
    pub updated_before: Option<chrono::DateTime<chrono::Utc>>,
    
    // Sorting and pagination
    pub sort_by: Option<String>,
    pub sort_order: Option<String>,
    pub limit: Option<u32>,
    pub offset: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Region {
    pub top_left: Position,
    pub bottom_right: Position,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QueryResult {
    pub blocks: Vec<UnifiedBlock>,
    pub total_count: u32,
    pub has_more: bool,
}

// ============================================================================
// ERROR TYPES
// ============================================================================

#[derive(Debug, Serialize, Deserialize)]
pub struct UnifiedError {
    pub code: String,
    pub message: String,
    pub details: Option<serde_json::Value>,
}

impl std::fmt::Display for UnifiedError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{}: {}", self.code, self.message)
    }
}

impl std::error::Error for UnifiedError {}

// ============================================================================
// OPERATION RESULT TYPES
// ============================================================================

#[derive(Debug, Serialize, Deserialize)]
pub struct BlockOperationResult {
    pub success: bool,
    pub block_id: Option<String>,
    pub error: Option<String>,
    pub affected_blocks: Option<Vec<String>>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct WorkspaceOperationResult {
    pub success: bool,
    pub workspace_id: Option<String>,
    pub error: Option<String>,
    pub blocks_loaded: Option<u32>,
    pub documents_loaded: Option<u32>,
    pub canvases_loaded: Option<u32>,
}
